HT-2827 - accept eduPersonEntitlement value of common-lib-terms

respinos opened the pull request:
- add value class to support poking at `$ENV{entitlements}`
- modify `Auth::Auth::affiliation_is_hathitrust` to pass if the user has `$ENTITLEMENT_COMMON_LIB_TERMS`
- added `get_shibboleth_entityID` to reject unexpected entityIDs

respinos committed:
run just the tests passed as arguments

respinos committed:
check that we have an institution before munging

respinos committed:
add Auth::Auth::Entitlement to support entitlements;
treat common-lib-terms as a valid affiliation;
ignore unknown entityids in get_shibboleth_entityID

aelkiss reviewed (commented):
Looks good to me at first glance. Is this running on a beta-* or somewhere else in dev I could test it?

respinos commented:
I've updated beta-3 with the mdp-lib changes.

aelkiss commented on t/access_common_lib_terms_entitlement.t:
```diff
@@ -0,0 +1,194 @@
+$ENV{entitlement} = q{urn:mace:dir:entitlement:common-lib-terms};
+
+# urn:mace:dir:entitlement:common-lib-terms from member institution
+foreach my $test ( @$member_tests ) {
```
I'm actually not sure why this is passing right now. I'll see if I can replicate locally...

aelkiss replied:
I can replicate it passing for me on grog, but this doesn't pass for me with docker -- it reports

```
/htapps/babel/mdp-lib/t/../t/access_common_lib_terms_entitlement.t .. 1/? ASSERT_FAIL: DBI error on statement=SELECT * FROM ht_institutions WHERE entityID=?: DBD::mysql::st execute failed: Table 'ht.ht_institutions' doesn't exist at /htapps/babel/mdp-lib/t/../DbUtils.pm line 133.
```

If we want that table to exist for tests, we could use https://github.com/hathitrust/db-image instead of the SQL schema included here.

We should probably set up CI for this repository with github actions -- should be straightforward to adapt from https://github.com/hathitrust/feed/blob/master/.github/workflows/ci.yml :+1: 
```
name: Run CI

on:
  push:
    branches:
      - master

  pull_request:

jobs:
  # Run tests
  test:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Check out submodules
        run: git submodule update --init metslib

      - name: Build docker image
        run: docker-compose build

      - name: Run tests
        run: docker-compose run test
```

aelkiss replied:
I'll look more tomorrow about why this is passing on grog as-is.

aelkiss replied:
Looking into this more, I think the reason this is passing is because it doesn't do anything to actually check the expected download status, and the access status doesn't differ between members and non-members.

respinos replied:
Indeed. Added tests for full-volume download.

respinos replied:
re: docker --- ah, that's interesting. I'll have to take a closer look at that. In the meanwhile, I added the non-member institution to the mocked hash which should prevent the lookup in `ht_institutions`

aelkiss replied:
:+1: Verified tests are passing in docker right now. I pushed a commit to set up CI for this.

aelkiss commented on Auth/Auth.pm:
```diff
@@ -859,6 +861,41 @@ sub get_eduPersonUnScopedAffiliation {
+    my $C = shift;
+
+    return Auth::Auth::Entitlement->new(
+        $self->has_expected_shibboleth_entityID($C) ? $ENV{entitlement} : ''
```
`$self->has_expected_shibboleth_entityID($C)` always returns false; it's expecting a second parameter of the actual entityID. Perhaps instead:

```perl
$self->get_shibboleth_entityID($C) ? $ENV{entitlement} : ''
```
```

respinos replied:
Updated to avoid extracting the entityid for this test.

aelkiss commented on Auth/Auth.pm:
```diff
@@ -1465,6 +1504,43 @@ sub is_cosign_active {
+    bless $self, $class;
+
+    if ( $entitlement ) {
+        $$self{entitlement_map} = map { $_ => 1 } split(/\s*;\s*/, $entitlement);
```
I think this ends up evaluating `map` in scalar context; try:

```perl
$$self{entitlement_map} = { map { $_ => 1 } split(/\s*;\s*/, $entitlement) };
```

respinos replied:
🤦

aelkiss reviewed (changes requested):
We should also perhaps change `debug=auth` maybe to:

```perl
     . q{ print-disabled-entitlement=} . $self->get_eduPersonEntitlement_print_disabled($C)
     . q{ computed-entitlements=} . join(',',@{$self->get_eduPersonEntitlement($C)->entitlements()})
``` 

I'll see if I can figure out why the test is unexpectedly passing.

respinos committed:
- Fix building of entitlement hash
- add entitlement to debug=auth
- add tests for Auth::Auth::Entitlement
- also check for full-volume download (the better difference between
member and not member)

respinos committed:
mock non-member provider

aelkiss committed:
Add CI with Github Actions

Docker layer caching might help here given the large number of packages
to be installed, but we'll first try it without and see how long it
takes.

aelkiss commented:
Working as expected on beta-3 for me with `debug=auth` -- AccessType gets set to ht_affiliate for a user with the common-lib-terms entitlement, an IdP in the ht_institutions table, and no matching affiliation; it stays as `ordinary-user` if the IdP is not in the ht_institutions table.
